hkmc2/shared/src/test/mlscript-compile/Rendering.mls

```diff
@@ -35,27 +35,350 @@ fun interleave(sep)(...args) =
+      parentIndent = if indent is null then null else indent.repeat(level)
+      itemIdent = if indent is null then null else indent.repeat(level + 1)
+    while done() is false do
+      push of ...next(items.length is 0, singleLineLength)
+    // Complete the single line length.
+    set singleLineLength += 2 // " }"
+    let
+      layoutMultiLine = () => symbols.start + "\n" + itemIdent + items.join(",\n" + itemIdent)
+        + "\n" + parentIndent + symbols.end
+      layoutSingleLine = () => symbols.start + " " + items.join(", ") + " " + symbols.end
+    if
+      items.length is 0 then symbols.empty
+      indent is null then layoutSingleLine()
+      hasMultiline then layoutMultiLine()
+      singleLineLength <= breakLength then layoutSingleLine()
+      // Try to break the parent object into many lines and check if it fits.
+      let lengthIfBreakParent = parentIndent.length + (singleLineLength - startPos + keyLength)
+      lengthIfBreakParent <= breakLength then layoutSingleLine()
+      else layoutMultiLine()
+  fun renderValue(arg, level, keyLength, startPos) = if
+    arg is // primitive values
+      undefined then "undefined"
+      null      then "null"
+      Str       then JSON.stringify(arg)
+      BigInt    then arg.toString() + "n"
+      Symbol    and
+        arg.description is undefined then "Symbol()"
+        else "Symbol(" + arg.description + ")"
+      Num       then arg.toString()
+      true      then "true"
+      false     then "false"
+    // Exclude all 
+    visitedObjects.has(arg) then visitedObjects.get(arg)
+    visitingObjects.has(arg) and
+      visitingObjects.get(arg) is
+        // We have already found this circular reference before
+        Int as index then "[Circular *" + index + "]"
+        // The first time we find a circular reference
+        else
+          let index = circularCounter
+          set circularCounter += 1
+          visitingObjects.set(arg, index)
+          "[Circular *" + index + "]"
+    else...
+    // Otherwise, add this object to the visited objects
+    // `null` is a marker for visited objects
+    visitingObjects.set(arg, null)
+    let rendered = if
+      arg is
+        Unit then "()"
+        Array then renderArray of arg, level, keyLength, startPos
+        Set then renderSet of arg, level, keyLength, startPos
+        Map then renderMap of arg, level, keyLength, startPos
+        WeakSet then "WeakSet { <items unknown> }"
+        WeakMap then "WeakMap { <items unknown> }"
+        Error then arg.name + ": " + arg.message
+        Function then
+          let desc = Object.getOwnPropertyDescriptor(arg, "prototype")
+          let head = if
+            // ES 2015 classes: `desc.writable` is false
+            desc is Object and desc.writable is false then
+              "[class " + (if arg.name is "" then "(anonymous)" else arg.name) + "]"
+            else
+              "[Function" + (if arg.name is "" then "" else " " + arg.name) + "]"
```

Comment:
Why isn't this
```suggestion
              "[function" + (if arg.name is "" then "" else " " + arg.name) + "]"
```

Reply:
This is fixed in efc5461.

Reply:
I'm wondering about the printing of such values. I think we should use a syntax that feels consistent with the rest of the lanfuage. `[function foo]` and `[class Bar] { ... }` make no sense in MLscript.

How about we just make them `fun foo` and `class Bar { ... }`?

Reply:
I agree on this. It looks more MLscripty.